### DISCUSSION TITLE
Nextjs API route with prisma resulting in internal server error

### DISCUSSION THREAD

**Discussion by LGmatrix13** [prisma/prisma#6378] [Q&A] [answered]
API route code:
```
import prisma from '../../../lib/prisma'

export default async function handle(req, res) {
  const posts = await prisma.post.findMany({
    take: 12,
    where: { published: true },
    include: { user: true },
  })
  res.json(posts)
}
```
Models:
```
model post {
  id Int @id @default(autoincrement())
  user String
  url String
  image String
}

model Account {
  id                 Int       @default(autoincrement()) @id
  compoundId         String    @unique @map(name: "compound_id")
  userId             Int       @map(name: "user_id")
  providerType       String    @map(name: "provider_type")
  providerId         String    @map(name: "provider_id")
  providerAccountId  String    @map(name: "provider_account_id")
  refreshToken       String?   @map(name: "refresh_token")
  accessToken        String?   @map(name: "access_token")
  accessTokenExpires DateTime? @map(name: "access_token_expires")
  createdAt          DateTime  @default(now()) @map(name: "created_at")
  updatedAt          DateTime  @default(now()) @map(name: "updated_at")

  @@index([providerAccountId], name: "providerAccountId")
  @@index([providerId], name: "providerId")
  @@index([userId], name: "userId")

  @@map(name: "accounts")
}

model Session {
  id           Int      @default(autoincrement()) @id
  userId       Int      @map(name: "user_id")
  expires      DateTime
  sessionToken String   @unique @map(name: "session_token")
  accessToken  String   @unique @map(name: "access_token")
  createdAt    DateTime @default(now()) @map(name: "created_at")
  updatedAt    DateTime @default(now()) @map(name: "updated_at")

  @@map(name: "sessions")
}

model User {
  id            Int       @default(autoincrement()) @id
  name          String?
  email         String?   @unique
  emailVerified DateTime? @map(name: "email_verified")
  image         String?
  createdAt     DateTime  @default(now()) @map(name: "created_at")
  updatedAt     DateTime  @default(now()) @map(name: "updated_at")

  @@map(name: "users")
}

model VerificationRequest {
  id         Int      @default(autoincrement()) @id
  identifier String
  token      String   @unique
  expires    DateTime
  createdAt  DateTime @default(now()) @map(name: "created_at")
  updatedAt  DateTime @default(now()) @map(name: "updated_at")

  @@map(name: "verification_requests")
}
```

Oddly, prisma studio is not working with my DB. The DB, however, is online and working.

**Comment by janpio** [comment]
What error are you getting in your app?
What error are you getting in Prisma Studio?
Does `npx prisma introspect` work?

**Reply by LGmatrix13** [reply]
So when I visit the route it says
`internal server error`

When I open prisma studio it just keeps loading without a response.

I’m away from my computer rn, but I try imputing that into my terminal.

**Reply by janpio** [reply]
> So when I visit the route it says
> `internal server error`

There should be way to get the actual error message, that is causing it to render an Internal Server Error. Console log files and so on. Only with the actual error we can understand what is happening.

**Reply by m-abdelwahab** [reply]
Looking at your schema, there is no relation between the `post` model and the `User` model.  So in your query, adding `include: { user: true },` will result in an error. 

Also the `post` model doesn't have a `published` field. So filtering will not work and result in an error.

I would change your schema to the following:

```prisma

model Post {
  id        Int     @id @default(autoincrement())
  published Boolean
  url String
  image String
  User   User? @relation(fields: [userId], references: [id])
  userId Int?
}

model User {
  id            Int       @id @default(autoincrement())
  name          String?
  email         String?   @unique
  emailVerified DateTime? @map(name: "email_verified")
  image         String?
  createdAt     DateTime  @default(now()) @map(name: "created_at")
  updatedAt     DateTime  @updatedAt @map(name: "updated_at")

  posts         Post[]

  @@map(name: "users")
}

```

This way, each post will be associated with a user and a user has many posts.

Also, the convention we follow is naming the model in [Pascal Case](https://techterms.com/definition/pascalcase#:~:text=PascalCase%20is%20a%20naming%20convention,in%20PascalCase%20is%20always%20capitalized.), so I would rename `post` to `Post` in your schema.

**Reply by LGmatrix13** [reply]
@m-abdelwahab thank you! I’ll be at my computer tomorrow and I’ll try it out!

**Comment by LGmatrix13** [comment, marked as answer]
Thanks, @m-abdelwahab your suggestions worked!
> Looking at your schema, there is no relation between the `post` model and the `User` model. So in your query, adding `include: { user: true },` will result in an error.
> 
> Also the `post` model doesn't have a `published` field. So filtering will not work and result in an error.
> 
> I would change your schema to the following:
> 
> ```prisma
> model Post {
>   id        Int     @id @default(autoincrement())
>   published Boolean
>   url String
>   image String
>   User   User? @relation(fields: [userId], references: [id])
>   userId Int?
> }
> 
> model User {
>   id            Int       @id @default(autoincrement())
>   name          String?
>   email         String?   @unique
>   emailVerified DateTime? @map(name: "email_verified")
>   image         String?
>   createdAt     DateTime  @default(now()) @map(name: "created_at")
>   updatedAt     DateTime  @updatedAt @map(name: "updated_at")
> 
>   posts         Post[]
> 
>   @@map(name: "users")
> }
> ```
> 
> This way, each post will be associated with a user and a user has many posts.
> 
> Also, the convention we follow is naming the model in [Pascal Case](https://techterms.com/definition/pascalcase#:~:text=PascalCase%20is%20a%20naming%20convention,in%20PascalCase%20is%20always%20capitalized.), so I would rename `post` to `Post` in your schema.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
